File: leetcode/1854/maximum_population_year.py
class Solution:
    def maximumPopulation(self, logs: List[List[int]]) -> int:
        max_pop = 0

        # Find max and min of birth year O(n)
        max_death = 0
        min_birth = 0
        for log in logs:
            max_death = max(max_death, log[1])
            min_birth = min(min_birth, log[0])

        # Create list in range max_birth - min_birth to count people alive
        pop_count = [0 for x in range(max_death-min_birth + 1)]

        # Loop through logs and add count
        for log in logs:
            birth_year = log[0] - min_birth
            death_year = log[1] - min_birth

            for i in range(birth_year, death_year):
                pop_count[i] = pop_count[i] + 1

        # Return max of list
        year = 0
        for i, count in enumerate(pop_count):
            if count > max_pop:
                max_pop = count
                year = i
        year = year + min_birth
        return year

File: leetcode/1854/maximum_population_year.cpp
class Solution {

public:
  int maximumPopulation(vector<vector<int>>& logs) {
    int maxDeath = 0;
    int minBirth = 0;

    for (int i = 0; i < logs.size(); i++) {
      maxDeath = max(maxDeath, logs[i][1]);
      minBirth = min(minBirth, logs[i][0]);
    }

    // Create vector with count
    vector<int> popCount(maxDeath - minBirth, 0);

    // Add popcount per log
    for (vector<int> log : logs) {
      for (int i = log[0]; i < log[1]; i++) {
        popCount[i] += 1;
      }; 
    };

    // Get maximum year
    int max_year = 0;
    int max_count = 0;
    for (int i = 0; i < popCount.size(); i++) {
      int curr_count = popCount[i];
      if (curr_count > max_count) {
        max_year = i;
        max_count = curr_count;
      };
    };

    return max_year + minBirth;
  };

  int max(int a, int b) {
    if (a > b) {
        return a;
    };
    return b;
  };

  int min(int a , int b) {
    if (a > b) {
        return b;
    };
    return a;
  };
};
